webhook/mutate/webhook.go

```diff
@@ -0,0 +1,223 @@
+	w.Write(mutatedBody)
+}
+
+// HandleMutateNoSgx is called when the sgx injection label is not set
+func HandleMutateNoSgx(w http.ResponseWriter, r *http.Request) {
+	log.Println("Handling mutate request, omitting sgx injection")
+	body := checkRequest(w, r)
+	if body == nil {
+		// Error was already written to w
+		return
+	}
+
+	// mutate the request and add sgx tolerations to pod
+	mutatedBody, err := mutate(body, false)
+	if err != nil {
+		http.Error(w, "unable to mutate request", http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Write(mutatedBody)
+}
+
+// mutate handles the creation of json patches for pods
+func mutate(body []byte, injectSgx bool) ([]byte, error) {
+	admReviewReq := v1.AdmissionReview{}
+	if err := json.Unmarshal(body, &admReviewReq); err != nil {
+		return nil, errors.New("invalid admission review")
+	}
+
+	if admReviewReq.Request == nil {
+		return nil, errors.New("empty admission request")
+	}
+
+	var pod corev1.Pod
+	if err := json.Unmarshal(admReviewReq.Request.Object.Raw, &pod); err != nil {
+		return nil, errors.New("invalid pod")
+	}
+
+	// admission response
+	admReviewResponse := v1.AdmissionReview{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "AdmissionReview",
+			APIVersion: "admission.k8s.io/v1",
+		},
+		Response: &v1.AdmissionResponse{
+			UID: admReviewReq.Request.UID,
+		},
+	}
+
+	pT := v1.PatchTypeJSONPatch
+	admReviewResponse.Response.PatchType = &pT
+
+	// get marble type from pod labels
+	marbleType := pod.Labels["marblerun.marbletype"]
+	// reject pod if label does not exist
+	if len(marbleType) == 0 {
+		admReviewResponse.Response.Allowed = false
+		admReviewResponse.Response.Result = &metav1.Status{
+			Status: "Rejected",
+			Reason: "Missing required label: [marblerun.marbletype]",
+		}
+		bytes, err := json.Marshal(admReviewResponse)
+		if err != nil {
+			return nil, errors.New("unable to marshal admission response")
+		}
+		return bytes, nil
+	}
+
+	// get namespace of pod
+	namespace := pod.Namespace
+	if len(namespace) == 0 {
+		namespace = "default"
+	}
+
+	newEnvVars := []corev1.EnvVar{
+		{
+			Name:  "EDG_MARBLE_COORDINATOR_ADDR",
+			Value: CoordAddr,
+		},
+		{
+			Name:  "EDG_MARBLE_TYPE",
+			Value: marbleType,
+		},
+		{
+			Name:  "EDG_MARBLE_DNS_NAMES",
+			Value: fmt.Sprintf("%s,%s.%s,%s.%s.svc.cluster.local", marbleType, marbleType, namespace, marbleType, namespace),
+		},
+		{
+			Name: "EDG_MARBLE_UUID_FILE",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					APIVersion: "v1",
+					FieldPath:  "metadata.uid",
+				},
+			},
+		},
+	}
+
+	var patch []map[string]interface{}
+
+	// create env variable patches for each container of the pod
+	for idx, container := range pod.Spec.Containers {
+		patch = append(patch, addEnvVar(container.Env, newEnvVars, fmt.Sprintf("/spec/containers/%d/env", idx))...)
+	}
+
+	// add sgx tolerations if enabled
+	if injectSgx {
+		patch = append(patch, map[string]interface{}{
```

Comment:
During testing this worked fine since the pods i started had some default tolerations set by kubernetes.
But in case a pod starts without any tolerations set i have added some functionality to check if any tolerations are present and if not create the first tolerations as an array, similar to the way the env variables are created